pkg/estimator/server/estimate.go

```diff
@@ -110,3 +111,100 @@ func (es *AccurateSchedulerEstimatorServer) nodeMaxAvailableReplica(node *framew
+	trace := utiltrace.New("Preempting", utiltrace.Field{Key: "namespacedName", Value: request.Preemptor})
+	defer trace.LogIfLong(100 * time.Millisecond)
+
+	snapShot := schedcache.NewEmptySnapshot()
+	if err := es.Cache.UpdateSnapshot(snapShot); err != nil {
+		return []pb.ObjectReference{}, err
+	}
+	trace.Step("Snapshotting estimator cache and node infos for preemption completed")
+
+	if snapShot.NumNodes() == 0 {
+		return []pb.ObjectReference{}, nil
+	}
+
+	candidateVictims, err := es.selectVictims(ctx, snapShot, request)
+	if err != nil {
+		return []pb.ObjectReference{}, err
+	}
+	trace.Step("Preemption completed")
+
+	return *candidateVictims, nil
+}
+
+func (es *AccurateSchedulerEstimatorServer) selectVictims(
+	ctx context.Context,
+	snapshot *schedcache.Snapshot,
+	request *pb.PreemptionRequest,
+) (*[]pb.ObjectReference, error) {
+	allNodes, err := snapshot.NodeInfos().List()
+	if err != nil {
+		return &[]pb.ObjectReference{}, err
+	}
+	var (
+		requirements     = request.ReplicaRequirements
+		requiredReplicas = request.NumReplicas
+		affinity         = nodeutil.GetRequiredNodeAffinity(requirements)
+		tolerations      []corev1.Toleration
+	)
+
+	if requirements.NodeClaim != nil {
+		tolerations = requirements.NodeClaim.Tolerations
+	}
+	candidateVictims, err := preemptee.ListCandidateVictims(request.Preemptor.Kind, &es.PreempteeLister)
+	if err != nil {
+		return &[]pb.ObjectReference{}, err
+	}
+	// run preemption plugins to filter out eligible victims for preemption
+	candidateVictims, ret := es.estimateFramework.RunPreemptionFilterPlugins(ctx, candidateVictims, request)
+
+	// No victims can be preempted on the cluster, skip further checks and return 0
+	if ret.IsUnschedulable() {
+		return &[]pb.ObjectReference{}, nil
+	}
+
+	if !ret.IsSuccess() && !ret.IsNoOperation() {
+		return &[]pb.ObjectReference{}, fmt.Errorf(fmt.Sprintf("preemption filter plugins fails with %s", ret.Reasons()))
+	}
+
+	orderedCandidateVictims, ret := es.estimateFramework.RunPreemptionOrderPlugins(ctx, candidateVictims, request)
+	if !ret.IsSuccess() && !ret.IsNoOperation() {
+		return &[]pb.ObjectReference{}, fmt.Errorf(fmt.Sprintf("preemption order plugins fails with %s", ret.Reasons()))
+	}
+	victimResourceBindings := make([]pb.ObjectReference, len(orderedCandidateVictims))
+	for _, candidateVictim := range orderedCandidateVictims {
+		victimResourceBindings = append(victimResourceBindings, *candidateVictim.ResourceBinding)
+	}
+	for cnt, candidateVictim := range orderedCandidateVictims {
+		var victimRbs int32
+		processNode := func(i int) {
+
+			node := allNodes[i]
+			if !nodeutil.IsNodeAffinityMatched(node.Node(), affinity) || !nodeutil.IsTolerationMatched(node.Node(), tolerations) {
+				return
+			}
+
+			nodeName := node.Node().Name
+			// Add each additional victim's resources to the releasedNodeResources
+			resource := candidateVictim.NodeResources[nodeName].ResourceList()
+			node.Allocatable.Add(resource)
```

Comment:
If you look at the node capacity
```
Capacity:
  cpu:                   <cpu-number>
  ephemeral-storage:     <ephemeral-storage>
  hugepages-1Gi:         <hugepages>
  memory:                <memory>
  nvidia.com/gpu:        <gpu>
  pods:                  110
```

There is a filed `pods` which indicates the max pod on a cluster is 110. Therefore, we also needs to track the the pod number on a Node.

For example, if the snapshot shows that the current node has 110 pods. Even after you add the resources back and find this node can fit a replica, but 

the rest.AllowedPodNumber will be zero because rest.AllowedPodNumber is 110, and node.Pods is 110
```
rest.AllowedPodNumber = util.MaxInt64(rest.AllowedPodNumber-int64(len(node.Pods)), 0)
```

If look at the MaxDivided function, it return the min between the divided results and r.AllowedPodNumber
```
res = MinInt64(res, r.AllowedPodNumber)
```